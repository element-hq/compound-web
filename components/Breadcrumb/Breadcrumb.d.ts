import { HTMLProps, JSX, MouseEventHandler } from '../../../node_modules/react';
interface BreadcrumbProps extends HTMLProps<HTMLElement> {
    /**
     * The label for the back button.
     */
    backLabel: string;
    /**
     * The click handler for the back button.
     */
    onBackClick: MouseEventHandler<HTMLButtonElement>;
    /**
     * The pages to display in the breadcrumb.
     * All the pages except the last one are displayed as links.
     */
    pages: string[];
    /**
     * The click handler for a page.
     * @param page - The page that was clicked.
     * @param index - The index of the page that was clicked.
     */
    onPageClick: (page: string, index: number) => void;
}
/**
 * A breadcrumb component.
 */
export declare function Breadcrumb({ backLabel, onBackClick, pages, onPageClick, className, ...props }: BreadcrumbProps): JSX.Element;
export {};
//# sourceMappingURL=Breadcrumb.d.ts.map