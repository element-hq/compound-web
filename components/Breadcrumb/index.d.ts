export { Breadcrumb } from './Breadcrumb';
//# sourceMappingURL=index.d.ts.map