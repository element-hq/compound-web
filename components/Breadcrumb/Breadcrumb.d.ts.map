{"version":3,"file":"Breadcrumb.d.ts","sourceRoot":"","sources":["../../../src/components/Breadcrumb/Breadcrumb.tsx"],"names":[],"mappings":"AAOA,OAAc,EAAE,SAAS,EAAE,GAAG,EAAE,iBAAiB,EAAiB,MAAM,OAAO,CAAC;AAOhF,UAAU,eAAgB,SAAQ,SAAS,CAAC,WAAW,CAAC;IACtD;;OAEG;IACH,SAAS,EAAE,MAAM,CAAC;IAClB;;OAEG;IACH,WAAW,EAAE,iBAAiB,CAAC,iBAAiB,CAAC,CAAC;IAClD;;;OAGG;IACH,KAAK,EAAE,MAAM,EAAE,CAAC;IAChB;;;;OAIG;IACH,WAAW,EAAE,CAAC,IAAI,EAAE,MAAM,EAAE,KAAK,EAAE,MAAM,KAAK,IAAI,CAAC;CACpD;AAED;;GAEG;AACH,wBAAgB,UAAU,CAAC,EACzB,SAAS,EACT,WAAW,EACX,KAAK,EACL,WAAW,EACX,SAAS,EACT,GAAG,KAAK,EACT,EAAE,eAAe,GAAG,GAAG,CAAC,OAAO,CAuB/B"}