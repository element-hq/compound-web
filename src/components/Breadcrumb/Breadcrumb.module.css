/*
Copyright 2024 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.breadcrumb {
  display: flex;
  align-items: center;
  block-size: 40px;
  gap: var(--cpd-space-3x);
  padding-block-end: var(--cpd-space-3x);
  border-block-end: 1px solid var(--cpd-color-alpha-gray-400);
  box-sizing: border-box;

  .pages {
    display: flex;
    gap: var(--cpd-space-1x);

    /* override list styles */
    list-style: none;
    margin: 0;
    padding: 0;

    a {
      cursor: pointer;
    }

    .last-page {
      font: var(--cpd-font-body-sm-regular);
      color: var(--cpd-color-text-secondary);
    }

    /*
     * Breadcrumb separator
     * We want this separator to be only visual and to not be in the accessibility tree.
     * The nav html element already provides an accessible way to separate the links.
     */
    li + li::before {
      display: inline-block;
      margin: 0 0.3em 0 0.25em;
      transform: rotate(15deg);
      border-inline-end: 1px solid var(--cpd-color-text-secondary);
      block-size: var(--cpd-space-3x);
      content: "";
    }

    /* Last page */
    :last-child {
      span {
        padding-inline-start: var(--cpd-space-1x);
      }
    }
  }
}
