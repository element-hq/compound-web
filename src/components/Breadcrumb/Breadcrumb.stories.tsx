/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { Breadcrumb as BreadcrumbComponent } from "./Breadcrumb";
import { Meta, StoryFn } from "@storybook/react-vite";
import React, { ComponentProps, useState } from "react";
import { Button } from "../Button";
import { fn } from "storybook/test";

export default {
  title: "Breadcrumb",
  component: BreadcrumbComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {
    pages: ["1st level page", "2nd level page", "Current page"],
    onPageClick: fn(),
    onBackClick: fn(),
    backLabel: "Back",
  },
} as Meta<typeof BreadcrumbComponent>;

const Template: StoryFn<typeof BreadcrumbComponent> = (
  args: ComponentProps<typeof BreadcrumbComponent>,
) => {
  return <BreadcrumbComponent {...args} />;
};

export const Default = Template.bind({});
Default.args = {};

function ControlledBreadcrumb(
  args: ComponentProps<typeof BreadcrumbComponent>,
) {
  const pagesContent = ["Page 1", "Page 2", "Page 3"];
  const [currentIndex, setCurrentIndex] = useState(2);
  const currentPage = pagesContent[currentIndex];

  return (
    <div
      style={{
        display: "flex",
        flexDirection: "column",
        gap: "8px",
        alignItems: "flex-start",
      }}
    >
      <BreadcrumbComponent
        backLabel="Back"
        pages={args.pages.slice(0, currentIndex + 1)}
        onPageClick={(_page, index) => setCurrentIndex(index)}
        onBackClick={() =>
          setCurrentIndex((_currentIndex) =>
            _currentIndex === 0 ? 0 : _currentIndex - 1,
          )
        }
      />
      {currentPage}
      <Button onClick={() => setCurrentIndex(2)}>Reset</Button>
    </div>
  );
}

export const Controlled: StoryFn<typeof BreadcrumbComponent> = (
  args: ComponentProps<typeof BreadcrumbComponent>,
) => {
  return <ControlledBreadcrumb {...args} />;
};
