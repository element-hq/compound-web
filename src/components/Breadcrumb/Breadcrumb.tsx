/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React, { HTMLProps, JSX, MouseEventHandler, KeyboardEvent } from "react";
import { IconButton } from "../Button";
import Chevron from "@vector-im/compound-design-tokens/assets/web/icons/chevron-left";
import styles from "./Breadcrumb.module.css";
import { Link } from "../Link/Link.tsx";
import classNames from "classnames";

interface BreadcrumbProps extends HTMLProps<HTMLElement> {
  /**
   * The label for the back button.
   */
  backLabel: string;
  /**
   * The click handler for the back button.
   */
  onBackClick: MouseEventHandler<HTMLButtonElement>;
  /**
   * The pages to display in the breadcrumb.
   * All the pages except the last one are displayed as links.
   */
  pages: string[];
  /**
   * The click handler for a page.
   * @param page - The page that was clicked.
   * @param index - The index of the page that was clicked.
   */
  onPageClick: (page: string, index: number) => void;
}

/**
 * A breadcrumb component.
 */
export function Breadcrumb({
  backLabel,
  onBackClick,
  pages,
  onPageClick,
  className,
  ...props
}: BreadcrumbProps): JSX.Element {
  return (
    <nav className={classNames(styles.breadcrumb, className)} {...props}>
      <IconButton
        kind="secondary"
        size="28px"
        aria-label={backLabel}
        onClick={onBackClick}
      >
        <Chevron />
      </IconButton>
      <ol className={styles.pages}>
        {pages.map((page, index) => (
          <Page
            key={index}
            page={page}
            isLastPage={index === pages.length - 1}
            onClick={() => onPageClick(page, index)}
          />
        ))}
      </ol>
    </nav>
  );
}

interface PageProps {
  /**
   * The page to display.
   */
  page: string;
  /**
   * Whether this is the last page in the breadcrumb.
   */
  isLastPage: boolean;
  /**
   * The click handler for the page, ignore for last page.
   */
  onClick: () => void;
}

/**
 * A breadcrumb page.
 * If not the last page, the page is displayed in a link.
 */
function Page({ page, isLastPage, onClick }: PageProps): JSX.Element {
  const onKeyDown = (event: KeyboardEvent<HTMLAnchorElement>) => {
    if (event.key === " ") {
      onClick();
    }
  };

  return (
    <li>
      {isLastPage ? (
        <span className={styles["last-page"]} aria-current="page">
          {page}
        </span>
      ) : (
        <Link
          size="small"
          role="button"
          onClick={onClick}
          onKeyDown={onKeyDown}
          tabIndex={0}
        >
          {page}
        </Link>
      )}
    </li>
  );
}
