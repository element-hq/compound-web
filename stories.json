{"v":3,"stories":{"actioncontrol--round":{"id":"actioncontrol--round","name":"Round","title":"ActionControl","importPath":"./src/components/ActionControl/ActionControl.stories.tsx","tags":["story"],"kind":"ActionControl","story":"Round","parameters":{"__id":"actioncontrol--round","docsOnly":false,"fileName":"./src/components/ActionControl/ActionControl.stories.tsx"}},"alert--docs":{"id":"alert--docs","title":"Alert","name":"Docs","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Alert","story":"Docs","parameters":{"__id":"alert--docs","docsOnly":true,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--success":{"id":"alert--success","name":"Success","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Success","parameters":{"__id":"alert--success","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--critical":{"id":"alert--critical","name":"Critical","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Critical","parameters":{"__id":"alert--critical","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--info":{"id":"alert--info","name":"Info","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Info","parameters":{"__id":"alert--info","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"alert--without-close":{"id":"alert--without-close","name":"Without Close","title":"Alert","importPath":"./src/components/Alert/Alert.stories.tsx","tags":["autodocs","story"],"kind":"Alert","story":"Without Close","parameters":{"__id":"alert--without-close","docsOnly":false,"fileName":"./src/components/Alert/Alert.stories.tsx"}},"avatar--docs":{"id":"avatar--docs","title":"Avatar","name":"Docs","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Avatar","story":"Docs","parameters":{"__id":"avatar--docs","docsOnly":true,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--round":{"id":"avatar--round","name":"Round","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Round","parameters":{"__id":"avatar--round","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--square":{"id":"avatar--square","name":"Square","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Square","parameters":{"__id":"avatar--square","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--button":{"id":"avatar--button","name":"Button","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Button","parameters":{"__id":"avatar--button","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--no-image-fallback":{"id":"avatar--no-image-fallback","name":"No Image Fallback","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"No Image Fallback","parameters":{"__id":"avatar--no-image-fallback","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--large-no-image-fallback":{"id":"avatar--large-no-image-fallback","name":"Large No Image Fallback","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"Large No Image Fallback","parameters":{"__id":"avatar--large-no-image-fallback","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar--all-avatars":{"id":"avatar--all-avatars","name":"All Avatars","title":"Avatar","importPath":"./src/components/Avatar/Avatar.stories.tsx","tags":["autodocs","story"],"kind":"Avatar","story":"All Avatars","parameters":{"__id":"avatar--all-avatars","docsOnly":false,"fileName":"./src/components/Avatar/Avatar.stories.tsx"}},"avatar-stackedavatar--docs":{"id":"avatar-stackedavatar--docs","title":"Avatar/StackedAvatar","name":"Docs","importPath":"./src/components/Avatar/AvatarStack.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Avatar/StackedAvatar","story":"Docs","parameters":{"__id":"avatar-stackedavatar--docs","docsOnly":true,"fileName":"./src/components/Avatar/AvatarStack.stories.tsx"}},"avatar-stackedavatar--avatar-stack":{"id":"avatar-stackedavatar--avatar-stack","name":"Avatar Stack","title":"Avatar/StackedAvatar","importPath":"./src/components/Avatar/AvatarStack.stories.tsx","tags":["autodocs","story"],"kind":"Avatar/StackedAvatar","story":"Avatar Stack","parameters":{"__id":"avatar-stackedavatar--avatar-stack","docsOnly":false,"fileName":"./src/components/Avatar/AvatarStack.stories.tsx"}},"badge--badge":{"id":"badge--badge","name":"Badge","title":"Badge","importPath":"./src/components/Badge/Badge.stories.tsx","tags":["story"],"kind":"Badge","story":"Badge","parameters":{"__id":"badge--badge","docsOnly":false,"fileName":"./src/components/Badge/Badge.stories.tsx"}},"button--docs":{"id":"button--docs","title":"Button","name":"Docs","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Button","story":"Docs","parameters":{"__id":"button--docs","docsOnly":true,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--primary":{"id":"button--primary","name":"Primary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Primary","parameters":{"__id":"button--primary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--secondary":{"id":"button--secondary","name":"Secondary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Secondary","parameters":{"__id":"button--secondary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--tertiary":{"id":"button--tertiary","name":"Tertiary","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Tertiary","parameters":{"__id":"button--tertiary","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"button--destructive":{"id":"button--destructive","name":"Destructive","title":"Button","importPath":"./src/components/Button/Button.stories.tsx","tags":["autodocs","story"],"kind":"Button","story":"Destructive","parameters":{"__id":"button--destructive","docsOnly":false,"fileName":"./src/components/Button/Button.stories.tsx"}},"checkbox--docs":{"id":"checkbox--docs","title":"Checkbox","name":"Docs","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Checkbox","story":"Docs","parameters":{"__id":"checkbox--docs","docsOnly":true,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--primary":{"id":"checkbox--primary","name":"Primary","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Primary","parameters":{"__id":"checkbox--primary","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--critical":{"id":"checkbox--critical","name":"Critical","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Critical","parameters":{"__id":"checkbox--critical","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--disabled":{"id":"checkbox--disabled","name":"Disabled","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Disabled","parameters":{"__id":"checkbox--disabled","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"checkbox--disabled-checked":{"id":"checkbox--disabled-checked","name":"Disabled Checked","title":"Checkbox","importPath":"./src/components/Checkbox/Checkbox.stories.tsx","tags":["autodocs","story"],"kind":"Checkbox","story":"Disabled Checked","parameters":{"__id":"checkbox--disabled-checked","docsOnly":false,"fileName":"./src/components/Checkbox/Checkbox.stories.tsx"}},"form--root":{"id":"form--root","name":"Root","title":"Form","importPath":"./src/components/Form/Form.stories.tsx","tags":["story"],"kind":"Form","story":"Root","parameters":{"__id":"form--root","docsOnly":false,"fileName":"./src/components/Form/Form.stories.tsx"}},"glass--docs":{"id":"glass--docs","title":"Glass","name":"Docs","importPath":"./src/components/Glass/Glass.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Glass","story":"Docs","parameters":{"__id":"glass--docs","docsOnly":true,"fileName":"./src/components/Glass/Glass.stories.tsx"}},"glass--glass":{"id":"glass--glass","name":"Glass","title":"Glass","importPath":"./src/components/Glass/Glass.stories.tsx","tags":["autodocs","story"],"kind":"Glass","story":"Glass","parameters":{"__id":"glass--glass","docsOnly":false,"fileName":"./src/components/Glass/Glass.stories.tsx"}},"icon--icon":{"id":"icon--icon","name":"Icon","title":"Icon","importPath":"./src/components/Icon/Icon.stories.tsx","tags":["story"],"kind":"Icon","story":"Icon","parameters":{"__id":"icon--icon","docsOnly":false,"fileName":"./src/components/Icon/Icon.stories.tsx"}},"button-iconbutton--docs":{"id":"button-iconbutton--docs","title":"Button/IconButton","name":"Docs","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Button/IconButton","story":"Docs","parameters":{"__id":"button-iconbutton--docs","docsOnly":true,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--normal":{"id":"button-iconbutton--normal","name":"Normal","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"Normal","parameters":{"__id":"button-iconbutton--normal","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--with-indicator":{"id":"button-iconbutton--with-indicator","name":"With Indicator","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"With Indicator","parameters":{"__id":"button-iconbutton--with-indicator","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"button-iconbutton--with-highlight-indicator":{"id":"button-iconbutton--with-highlight-indicator","name":"With Highlight Indicator","title":"Button/IconButton","importPath":"./src/components/IconButton/IconButton.stories.tsx","tags":["autodocs","story"],"kind":"Button/IconButton","story":"With Highlight Indicator","parameters":{"__id":"button-iconbutton--with-highlight-indicator","docsOnly":false,"fileName":"./src/components/IconButton/IconButton.stories.tsx"}},"link--docs":{"id":"link--docs","title":"Link","name":"Docs","importPath":"./src/components/Link/Link.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Link","story":"Docs","parameters":{"__id":"link--docs","docsOnly":true,"fileName":"./src/components/Link/Link.stories.tsx"}},"link--round":{"id":"link--round","name":"Round","title":"Link","importPath":"./src/components/Link/Link.stories.tsx","tags":["autodocs","story"],"kind":"Link","story":"Round","parameters":{"__id":"link--round","docsOnly":false,"fileName":"./src/components/Link/Link.stories.tsx"}},"menuitem--docs":{"id":"menuitem--docs","title":"MenuItem","name":"Docs","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"MenuItem","story":"Docs","parameters":{"__id":"menuitem--docs","docsOnly":true,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"menuitem--primary":{"id":"menuitem--primary","name":"Primary","title":"MenuItem","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","story"],"kind":"MenuItem","story":"Primary","parameters":{"__id":"menuitem--primary","docsOnly":false,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"menuitem--critical":{"id":"menuitem--critical","name":"Critical","title":"MenuItem","importPath":"./src/components/MenuItem/MenuItem.stories.tsx","tags":["autodocs","story"],"kind":"MenuItem","story":"Critical","parameters":{"__id":"menuitem--critical","docsOnly":false,"fileName":"./src/components/MenuItem/MenuItem.stories.tsx"}},"togglemenuitem--docs":{"id":"togglemenuitem--docs","title":"ToggleMenuItem","name":"Docs","importPath":"./src/components/MenuItem/ToggleMenuItem.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"ToggleMenuItem","story":"Docs","parameters":{"__id":"togglemenuitem--docs","docsOnly":true,"fileName":"./src/components/MenuItem/ToggleMenuItem.stories.tsx"}},"togglemenuitem--primary":{"id":"togglemenuitem--primary","name":"Primary","title":"ToggleMenuItem","importPath":"./src/components/MenuItem/ToggleMenuItem.stories.tsx","tags":["autodocs","story"],"kind":"ToggleMenuItem","story":"Primary","parameters":{"__id":"togglemenuitem--primary","docsOnly":false,"fileName":"./src/components/MenuItem/ToggleMenuItem.stories.tsx"}},"radio--docs":{"id":"radio--docs","title":"Radio","name":"Docs","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Radio","story":"Docs","parameters":{"__id":"radio--docs","docsOnly":true,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--primary":{"id":"radio--primary","name":"Primary","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Primary","parameters":{"__id":"radio--primary","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--checked":{"id":"radio--checked","name":"Checked","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Checked","parameters":{"__id":"radio--checked","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--checked-disabled":{"id":"radio--checked-disabled","name":"Checked Disabled","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Checked Disabled","parameters":{"__id":"radio--checked-disabled","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical":{"id":"radio--critical","name":"Critical","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical","parameters":{"__id":"radio--critical","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical-checked":{"id":"radio--critical-checked","name":"Critical Checked","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical Checked","parameters":{"__id":"radio--critical-checked","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"radio--critical-checked-disabled":{"id":"radio--critical-checked-disabled","name":"Critical Checked Disabled","title":"Radio","importPath":"./src/components/Radio/Radio.stories.tsx","tags":["autodocs","story"],"kind":"Radio","story":"Critical Checked Disabled","parameters":{"__id":"radio--critical-checked-disabled","docsOnly":false,"fileName":"./src/components/Radio/Radio.stories.tsx"}},"search--docs":{"id":"search--docs","title":"Search","name":"Docs","importPath":"./src/components/Search/Search.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Search","story":"Docs","parameters":{"__id":"search--docs","docsOnly":true,"fileName":"./src/components/Search/Search.stories.tsx"}},"search--search":{"id":"search--search","name":"Search","title":"Search","importPath":"./src/components/Search/Search.stories.tsx","tags":["autodocs","story"],"kind":"Search","story":"Search","parameters":{"__id":"search--search","docsOnly":false,"fileName":"./src/components/Search/Search.stories.tsx"}},"toggle--docs":{"id":"toggle--docs","title":"Toggle","name":"Docs","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Toggle","story":"Docs","parameters":{"__id":"toggle--docs","docsOnly":true,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--active":{"id":"toggle--active","name":"Active","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Active","parameters":{"__id":"toggle--active","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--checked":{"id":"toggle--checked","name":"Checked","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Checked","parameters":{"__id":"toggle--checked","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--disabled":{"id":"toggle--disabled","name":"Disabled","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Disabled","parameters":{"__id":"toggle--disabled","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"toggle--disabled-checked":{"id":"toggle--disabled-checked","name":"Disabled Checked","title":"Toggle","importPath":"./src/components/Toggle/Toggle.stories.tsx","tags":["autodocs","story"],"kind":"Toggle","story":"Disabled Checked","parameters":{"__id":"toggle--disabled-checked","docsOnly":false,"fileName":"./src/components/Toggle/Toggle.stories.tsx"}},"tooltip--docs":{"id":"tooltip--docs","title":"Tooltip","name":"Docs","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","docs"],"storiesImports":[],"kind":"Tooltip","story":"Docs","parameters":{"__id":"tooltip--docs","docsOnly":true,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"tooltip--side":{"id":"tooltip--side","name":"Side","title":"Tooltip","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","story"],"kind":"Tooltip","story":"Side","parameters":{"__id":"tooltip--side","docsOnly":false,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"tooltip--align":{"id":"tooltip--align","name":"Align","title":"Tooltip","importPath":"./src/components/Tooltip/Tooltip.stories.tsx","tags":["autodocs","story"],"kind":"Tooltip","story":"Align","parameters":{"__id":"tooltip--align","docsOnly":false,"fileName":"./src/components/Tooltip/Tooltip.stories.tsx"}},"typography--heading":{"id":"typography--heading","name":"Heading","title":"Typography","importPath":"./src/components/Typography/Heading.stories.tsx","tags":["story"],"kind":"Typography","story":"Heading","parameters":{"__id":"typography--heading","docsOnly":false,"fileName":"./src/components/Typography/Heading.stories.tsx"}},"typography--text":{"id":"typography--text","name":"Text","title":"Typography","importPath":"./src/components/Typography/Text.stories.tsx","tags":["story"],"kind":"Typography","story":"Text","parameters":{"__id":"typography--text","docsOnly":false,"fileName":"./src/components/Typography/Text.stories.tsx"}}}}
