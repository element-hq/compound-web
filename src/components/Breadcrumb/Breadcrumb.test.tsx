/*
Copyright 2024 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import { describe, expect, it, vi } from "vitest";
import { composeStories } from "@storybook/react";
import * as stories from "./Breadcrumb.stories.tsx";
import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { Breadcrumb } from "./Breadcrumb.tsx";

const { Default } = composeStories(stories);

describe("Breadcrumb", () => {
  it("should render", () => {
    render(<Default />);
    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  it("should call onPageClick when a page is clicked", async () => {
    const user = userEvent.setup();
    const onPageClick = vi.fn();
    const onBackClick = vi.fn();

    render(
      <Breadcrumb
        pages={["1st level page", "2nd level page", "Current page"]}
        onPageClick={onPageClick}
        onBackClick={onBackClick}
        backLabel="Back"
      />,
    );

    // Click listener
    await user.click(screen.getByRole("button", { name: "1st level page" }));
    expect(onPageClick).toHaveBeenCalledWith("1st level page", 0);

    onPageClick.mockReset();
    // Keyboard listener
    await user.type(
      screen.getByRole("button", { name: "1st level page" }),
      " ",
    );
    expect(onPageClick).toHaveBeenCalledWith("1st level page", 0);

    // Back button
    await user.click(screen.getByRole("button", { name: "Back" }));
    expect(onBackClick).toHaveBeenCalled();
  });
});
