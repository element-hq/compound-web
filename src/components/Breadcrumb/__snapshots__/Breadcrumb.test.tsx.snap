// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb > should render 1`] = `
<nav
  class="_breadcrumb_c693d3"
>
  <button
    aria-label="Back"
    class="_icon-button_e3253e"
    data-kind="secondary"
    role="button"
    style="--cpd-icon-button-size: 28px;"
    tabindex="0"
  >
    <div
      class="_indicator-icon_ab1fc6"
      style="--cpd-icon-button-size: 100%;"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m13.3 17.3-4.6-4.6a.9.9 0 0 1-.213-.325A1.1 1.1 0 0 1 8.425 12q0-.2.062-.375A.9.9 0 0 1 8.7 11.3l4.6-4.6a.95.95 0 0 1 .7-.275q.425 0 .7.275a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7L10.8 12l3.9 3.9a.95.95 0 0 1 .275.7.95.95 0 0 1-.275.7.95.95 0 0 1-.7.275.95.95 0 0 1-.7-.275"
        />
      </svg>
    </div>
  </button>
  <ol
    class="_pages_c693d3"
  >
    <li>
      <a
        class="_link_379d57"
        data-kind="primary"
        data-size="small"
        rel="noreferrer noopener"
        role="button"
        tabindex="0"
      >
        1st level page
      </a>
    </li>
    <li>
      <a
        class="_link_379d57"
        data-kind="primary"
        data-size="small"
        rel="noreferrer noopener"
        role="button"
        tabindex="0"
      >
        2nd level page
      </a>
    </li>
    <li>
      <span
        aria-current="page"
        class="_last-page_c693d3"
      >
        Current page
      </span>
    </li>
  </ol>
</nav>
`;
